### PR TITLE
dnsdist: Fix warnings reported by clang's analyzer and cppcheck

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Mostly non-noticeable performance hints:
- pass shared pointers by reference whenever possible
- use pre-increment instead of post-increment for non-primitive types
- use initialization list instead of assigning values in the constructor's body

`LuaWrapper` moving the function variable before using it again might be a real issue though, and the locks' move constructor might have become one at some point.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled and tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdist-lua.hh [hunk 1]
```diff
@@ -25,7 +25,7 @@ class LuaAction : public DNSAction
 {
 public:
   typedef std::function<std::tuple<int, boost::optional<string> >(DNSQuestion* dq)> func_t;
-  LuaAction(LuaAction::func_t func) : d_func(func)
+  LuaAction(LuaAction::func_t& func) : d_func(func)
```
Comment (chbruyand) [inline pdns/dnsdist-lua.hh]: Maybe `const LuaAction::func_t& func` ?

File: pdns/dnsdist-lua.hh [hunk 2]
```diff
@@ -40,7 +40,7 @@ class LuaResponseAction : public DNSResponseAction
 {
 public:
   typedef std::function<std::tuple<int, boost::optional<string> >(DNSResponse* dr)> func_t;
-  LuaResponseAction(LuaResponseAction::func_t func) : d_func(func)
+  LuaResponseAction(LuaResponseAction::func_t& func) : d_func(func)
```
Comment (chbruyand) [inline pdns/dnsdist-lua.hh]: Maybe `const LuaResponseAction::func_t& func` ?

File: pdns/dnsdist.hh [hunk 3]
```diff
@@ -324,12 +337,10 @@ struct ClientState;
 struct IDState
 {
   IDState() : origFD(-1), sentTime(true), delayMsec(0), tempFailureTTL(boost::none) { origDest.sin4.sin_family = 0;}
-  IDState(const IDState& orig)
+  IDState(const IDState& orig): origRemote(orig.origRemote), origDest(orig.origDest)
```
Comment (chbruyand) [inline pdns/dnsdist.hh]: Why not also moving other member variables in initialization list ? (`origFD`, `origID`, `delayMsec` and `tempFailureTTL`)

File: pdns/dnsparser.cc [hunk 4]
```diff
@@ -199,9 +199,8 @@ DNSRecordContent::zmakermap_t& DNSRecordContent::getZmakermap()
   return zmakermap;
 }
 
-DNSRecord::DNSRecord(const DNSResourceRecord& rr)
+DNSRecord::DNSRecord(const DNSResourceRecord& rr): d_name(rr.qname)
 {
-  d_name = rr.qname;
   d_type = rr.qtype.getCode();
   d_ttl = rr.ttl;
```
Comment (chbruyand) [inline pdns/dnsparser.cc]: Same remark as with `IDState()`. But I may be missing something. Sorry if that's the case.

**Comment by rgacogne** [comment]
Pushed a fix to constify the function references passed to `LuaAction` and `LuaResponseAction`.

File: pdns/dnsdist.hh (hunk 3 above)
Comment (rgacogne) [reply pdns/dnsdist.hh]: I had to keep the initialization of some members in the constructor body (copy constructor is missing for `std::atomic` for example), so I only changed the one that mattered, ie non-Plain Old Data types.

File: pdns/dnsparser.cc (hunk 4 above)
Comment (rgacogne) [reply pdns/dnsparser.cc]: Same reason here.

**Pull request merged**